prboom2/src/gl_main.c

```diff
@@ -2328,20 +2328,26 @@ static void gld_DrawSprite(GLSprite *sprite)
 
 static void gld_AddHealthBar(mobj_t* thing, GLSprite *sprite)
 {
-  if (((thing->flags & (MF_COUNTKILL | MF_CORPSE)) == MF_COUNTKILL) && (thing->health > 0))
+  if ((thing->flags & MF_SHOOTABLE) && thing->info->spawnhealth > 0 &&
+      thing->health > 0)
   {
     GLHealthBar hbar;
     int health_percent = thing->health * 100 / thing->info->spawnhealth;
 
-    hbar.cm = -1;
-    if (health_percent <= health_bar_red)
-      hbar.cm = CR_RED;
-    else if (health_percent <= health_bar_yellow)
-      hbar.cm = CR_YELLOW;
-    else if (health_percent <= health_bar_green)
-      hbar.cm = CR_GREEN;
+    if (health_percent <= health_bar_red) {
+      hbar.r = 1;
+      hbar.g = 0;
+    } else if (health_percent <= health_bar_yellow) {
+      hbar.r = 1;
+      hbar.g = (double)(health_percent - health_bar_red) /
+               (health_bar_yellow - health_bar_red);
+    } else if (health_percent <= health_bar_green) {
+      /* FullSimplify[1-((percent-yellow)/(green-yellow))] */
+      hbar.r = (double)(health_bar_green - health_percent) /
+               (health_bar_green - health_bar_yellow);
+      hbar.g = 1;
+    }
 
-    if (hbar.cm >= 0)
     {
```

Comment:
Is this in a code block on purpose?

Reply:
I did this intentionally, yes. By showing the health bar at all times, one becomes aware that something is shootable in the first place, like customized decoration. Hence the (hbar.cm >= 0) condition was removed. As new variables are declared and C89 is picky about mixing declarations and code, I had just left the block there to get a minimal patch. I can certainly add a patch to remedy that.

Having bars everywhere is not so much a detriment; in case of e.g. nuts.wad, most monsters will get damaged anyway so that even the original health bar's implementation would readily show across the rendered scene.